src/deploy.ts

```diff
@@ -86,51 +102,121 @@ export class HypERC20Deployer extends GasRouterDeployer<
+  ): Promise<HypERC20> {
+    this.logger(`Deploying synthetic router on ${chain}...`);
+    const router = await this.deployContractFromFactory(
+      chain,
+      new HypERC20__factory(),
+      'HypERC20',
+      [tokenMetadata.decimals],
+    );
+    await this.multiProvider.handleTx(
+      chain,
+      router.initialize(
+        config.mailbox,
+        config.interchainGasPaymaster,
+        tokenMetadata.totalSupply,
+        tokenMetadata.name,
+        tokenMetadata.symbol,
+      ),
+    );
+    return router;
+  }
+
+  async deploy(
+    partialDeployment?: ChainMap<HypERC20Contracts>,
+  ): Promise<ChainMap<HypERC20Contracts>> {
+    // deploy collateral or native first to populate token metadata for synthetics
+    for (const [chain, config] of Object.entries(this.configMap)) {
+      if (isCollateralConfig(config)) {
+        this.deployedContracts[chain] = {
+          router: await this.deployCollateral(chain as ChainName, config)
+        };
+      } else if (isNativeConfig(config)) {
+        this.deployedContracts[chain] = {
+          router: await this.deployNative(chain as ChainName, config)
+        };
+      }
+    }
+
+    // deploy synthetics
+    return super.deploy({ ...partialDeployment, ...this.deployedContracts });
+  }
+
   async deployContracts(chain: ChainName, config: HypERC20Config) {
-    if (isCollateralConfig(config)) {
-      const router = await this.deployContractFromFactory(
-        chain,
-        new HypERC20Collateral__factory(),
-        'HypERC20Collateral',
-        [config.token],
-      );
-      await this.multiProvider.handleTx(
-        chain,
-        router.initialize(config.mailbox, config.interchainGasPaymaster),
-      );
-      return { router };
-    } else if (isSyntheticConfig(config)) {
-      const router = await this.deployContractFromFactory(
-        chain,
-        new HypERC20__factory(),
-        'HypERC20',
-        [],
-      );
-      await this.multiProvider.handleTx(
-        chain,
-        router.initialize(
-          config.mailbox,
-          config.interchainGasPaymaster,
-          config.totalSupply,
-          config.name,
-          config.symbol,
-        ),
-      );
-      return { router };
-    } else if (isNativeConfig(config)) {
-      const router = await this.deployContractFromFactory(
-        chain,
-        new HypNative__factory(),
-        'HypNative',
-        [],
-      );
-      await this.multiProvider.handleTx(
-        chain,
-        router.initialize(config.mailbox, config.interchainGasPaymaster),
-      );
-      return { router };
+    // skip if already deployed
+    if (this.deployedContracts[chain]) {
+      return this.deployedContracts[chain];
     }
-    throw new Error('Invalid config');
+
+    if (!isSyntheticConfig(config)) {
+      throw new Error('Expect only synthetic configs');
+    }
+
+    const erc20Metadata = {...this.tokenMetadata, ...config };
+    if (!isErc20Metadata(erc20Metadata)) {
```

Comment:
Config validation needs to happen imo sooner, but maybe something that can be done outside. If you are not going to do this now, would you mind creating a ticket for this and assigning it to @jmrossy for this sprint?

Reply:
I can create a `Zod` schema for the configs and validate them at read-time if we want. Please do create a ticket if that's the way we want to go.

Regardless, keeping these checks in here as backup is still a good idea IMO.